src/com/ms/silverking/cloud/dht/daemon/storage/FileSegment.java

```diff
@@ -23,243 +23,255 @@
+      Log.warningf("raFile.length() %d", raFile.length());
+      Log.warningf("raFile.length() - dataSegmentSize %d", raFile.length() - dataSegmentSize);
     }
+    if (segmentIndexLocation == SegmentIndexLocation.RAM) {
+      byte[] _htBufArray;
 
-    // called from Create
-    private FileSegment(File nsDir, int segmentNumber, AccessMode accessMode, RandomAccessFile raFile, 
-            ByteBuffer dataBuf, int dataSegmentSize, NamespaceOptions nsOptions) 
-                    throws IOException {
-        super(nsDir, segmentNumber, dataBuf, StoreConfiguration.fileInitialCuckooConfig, dataSegmentSize, nsOptions);
-        this.accessMode = accessMode;
-        this.raFile = raFile;
-        this.storageFormat = FileSegmentStorageFormat.parse(nsOptions.getStorageFormat());
+      _htBufArray = new byte[(int) (raFile.length() - dataSegmentSize)];
+      raFile.seek(dataSegmentSize);
+      raFile.read(_htBufArray);
+      rawHTBuf = ByteBuffer.wrap(_htBufArray);
+    } else {
+      rawHTBuf = raFile.getChannel().map(MapMode.READ_ONLY, dataSegmentSize, raFile.length() - dataSegmentSize);
+      if (segmentPrereadMode == SegmentPrereadMode.Preread) {
+        ((MappedByteBuffer) rawHTBuf).load();
+      }
     }
-    
-    public void persist() throws IOException {
-        ((MappedByteBuffer)dataBuf).force();
-        FileSegmentMetaData.persist(storageFormat, raFile, dataSegmentSize, (IntArrayCuckoo)keyToOffset, (RAMOffsetListStore)offsetListStore, invalidatedOffsets);
-        raFile.close();
-        raFile = null;
-        close();
+    rawHTBuf = rawHTBuf.order(ByteOrder.nativeOrder());
+
+    if (debugMetaData) {
+      Log.warningf("rawHTBuf %s", rawHTBuf);
     }
-    
-    public void close() {
-        dataBuf = null;
+    fsm = FileSegmentMetaData.create(rawHTBuf, FileSegmentStorageFormat.parse(nsOptions.getStorageFormat()));
+    ktome = (KeyToOffsetMapElement) fsm.getElement(FSMElementType.OffsetMap);
+    //((IntBufferCuckoo)ktome.getKeyToOffsetMap()).display();
+    ole = (OffsetListsElement) fsm.getElement(FSMElementType.OffsetLists);
+    //((BufferOffsetListStore)ole.getOffsetListStore(nsOptions)).displayForDebug();
+    ioe = (InvalidatedOffsetsElement) fsm.getElement(FSMElementType.InvalidatedOffsets);
+    return new FileSegment(nsDir, segmentNumber, accessMode, raFile, dataBuf, ktome.getKeyToOffsetMap(),
+        ioe != null ? ioe.getInvalidatedOffsets() : null, ole.getOffsetListStore(nsOptions), dataSegmentSize,
+        nsOptions);
+  }
+
+  static File fileForSegment(File nsDir, int segmentNumber) {
+    return new File(nsDir, Integer.toString(segmentNumber));
+  }
+
+  public static ByteBuffer getDataSegment(File nsDir, int segmentNumber, int dataSegmentSize) throws IOException {
+    RandomAccessFile raFile;
+    ByteBuffer dataBuf;
+
+    raFile = new RandomAccessFile(fileForSegment(nsDir, segmentNumber), roFileMode);
+    dataBuf = raFile.getChannel().map(MapMode.READ_ONLY, 0, dataSegmentSize);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getChannel*. The resource is referenced in statements at the following line: 202. The resource closure statement is at line: 203. There are other execution paths that do not contain closure statements, for example, when FileChannel.map throws an exception. Either a) close the object returned by *getChannel()* in a try-finally block or b) close the resource by declaring the object returned by *getChannel()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).